### PR TITLE
feat: utxo validate addresses at Tx build time

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Does what it says on the box - screens at Tx build time instead of broadcast time so we can leverage UTXO inputs and outputs to screen *all* addresses involved in a UTXO Tx.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes N/A

None?

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Send an UTXO Tx e.g send with network tab open
- Notice many `screening` requests
- Open the explorer Tx link and notice inputs/outputs addresses match our `screening` calls

### Engineering

- ☝🏽 

### Operations

- ☝🏽 

## Screenshots (if applicable)

image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
* **#6510** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6510?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gomesalexandre and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/chain-adapters/src/utils/validateAddress.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ const TRM_LABS_API_URL = 'https://api.trmlabs.com/public/v1/sanctions/screening'
 
 const cache: Record<string, Promise<boolean>> = {}
 
-const _validateAddress = async (address: string): Promise<boolean> => {
+const checkIsSanctioned = async (address: string): Promise<boolean> => {
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utils/validateAddress.ts:7]: Renaming to make things clearer here, not related to this PR per se

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 2]
```diff
@@ -301,6 +301,26 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
...
+        address: account.chainSpecific.addresses?.[nextReceiveAddressIndex]?.pubkey,
+      }
+
+      const addresses = [...inputs, ...outputs, nextChangeAddressInput, nextReceiveAddressInput]
+        .map(({ address }) => address)
+        .filter(Boolean) as string[]
+
+      const uniqueAddresses = [...new Set(addresses)]
+
+      await Promise.all(uniqueAddresses.map(validateAddress))
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:322]: TRM can take multiple addresses as input so we could refactor the checks to be performed over a response containing multiple addresses, but not for this PR

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 3]
```diff
@@ -434,10 +453,7 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
     receiverAddress,
     signTxInput,
   }: SignAndBroadcastTransactionInput<T>): Promise<string> {
-    await Promise.all([
-      validateAddress(senderAddress),
-      receiverAddress !== CONTRACT_INTERACTION && validateAddress(receiverAddress),
-    ])
+    await (receiverAddress !== CONTRACT_INTERACTION && validateAddress(receiverAddress))
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:456]: Keeps only the receive side check now - the naming is odd but this refers (or at least *should* refer to) the *final* recipient e.g for a cross-chain swap, possibly a receive address on a diff chain

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 4]
```diff
@@ -479,15 +495,8 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
...
-    senderAddress,
-    receiverAddress,
-    hex,
-  }: BroadcastTransactionInput): Promise<string> {
-    await Promise.all([
-      validateAddress(senderAddress),
-      receiverAddress !== CONTRACT_INTERACTION && validateAddress(receiverAddress),
-    ])
+  async broadcastTransaction({ receiverAddress, hex }: BroadcastTransactionInput): Promise<string> {
+    await (receiverAddress !== CONTRACT_INTERACTION && validateAddress(receiverAddress))
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:499]: Keeps only the receive side check now - the naming is odd but this refers (or at least *should* refer to) the *final* recipient e.g for a cross-chain swap, possibly a receive address on a diff chain

**Review by woodenfurniture — COMMENTED** [review]
Tested with a BTC send, seeing multiple successful screening checks.

**Review by woodenfurniture — APPROVED** [review]

**Pull request merged**